### ISSUE TITLE
[Reliability Improvement] Record the step # in the optimizer state of PostLocalSGDOptimizer

### ISSUE THREAD

**Issue by wayi1** [pytorch/pytorch#74556]
### 🚀 The feature, motivation and pitch

**Context:**
[PostLocalSGDOptimizer](https://github.com/pytorch/pytorch/blob/8b0847811501cd452b131b6153f28c254a3ee44f/torch/distributed/optim/post_localSGD_optimizer.py#L5) can have a non-trivial warm-up stage. E.g., a training may take 30 epochs, and the warmup stage may take 3-5 epochs.

**Issue:**
If the training process is not very reliable and may fail a few times, then the training needs to resume from checkpoints. However, in the current setup the built-in [step counter](https://github.com/pytorch/pytorch/blob/8b8fac91bfca6cadf86d402622f31e8e2dcdc429/torch/distributed/algorithms/model_averaging/averagers.py#L22) in `ModelAverager` will be reset after each restart. As a result, the warm-up stage will run multiple times, which will lead to less speedup.

**Proposed Solution:**
To be able to skip the warm-up stage after resuming from a checkpiont, the step # across epochs must recorded and stored in the checkpoint as well. A good option is the optimizer state.

### Alternatives

If PyTorch Ignite or PyTorch Lightning is used, the global step # will be automatically stored in the checkpoint, and such feature can also be easily supported by modifying the application code.

However, it will be nice to support this feature in PyTorch directly, especially if an optimizer state is already available in this case.

**Comment by wayi1** [comment]
Another relevant proposal: https://github.com/pytorch/pytorch/issues/75666, inspired by this post: https://discuss.pytorch.org/t/how-to-resume-with-powersgd-enabled-training/148747

This is because this optimizer needs to be combined with post localSGD hook.

**Comment by rohan-varma** [comment]
reserved for intern ramp up

**Comment by awgu** [comment]
@wayi1 
Does `PostLocalSGDOptimizer` support or plan to support warming up at the parameter granularity? In other words, can some parameters have their `requires_grad` set to `False`, de-synchronizing their logical `step` values from the `step` values of the parameters with `requires_grad=True`?

We are wondering if if we save a `step` counter to the optimizer state whether it should be saved per parameter (e.g. `optim_state_dict["state"][param_id]["step"] = ...`) or globally to the optimizer (e.g. `optim_state_dict["step"] = ...`).

**Comment by wayi1** [comment]
> @wayi1 Does `PostLocalSGDOptimizer` support or plan to support warming up at the parameter granularity? In other words, can some parameters have their `requires_grad` set to `False`, de-synchronizing their logical `step` values from the `step` values of the parameters with `requires_grad=True`, meaning that they may need additional iterations later to warm up?
> 
> We are wondering if we save a `step` counter to the optimizer state whether it should be saved per parameter (e.g. `optim_state_dict["state"][param_id]["step"] = ...`) or globally to the optimizer (e.g. `optim_state_dict["step"] = ...`).

Sorry for the late reply.

`PostLocalSGDOptimizer` is only applied to the parameters that need to compute gradients. See: https://github.com/pytorch/pytorch/blob/da3c848dfa5e8269b4cd7da91ba32025b268f1d8/torch/distributed/algorithms/model_averaging/utils.py#L58

Therefore, it works well for my cases where each parameter is only updated by a single optimizer.
I understand that
